$$ ITI \space AI-Pro: \space Intake \space 45 $$
$$ Recommender \space Systems $$
$$ Lab \space no. \space 3 $$

# `01` Import Necessary Libraries

## `i` Default Libraries

In [ ]:
import numpy as np
import pandas as pd

## `ii` Additional Libraries
Add imports for additional libraries you used throughout the notebook

----------------------------

# `02` Load Data

Load `songsDataset.csv` file into a dataframe

In [ ]:
data = pd.read_csv("Data/songsDataset.csv")
data.head()

userID  songID  rating
0       0   90409       5
1       4   91266       1
2       5    8063       2
3       5   24427       4
4       5  105433       4

--------------------------

# `03` Matrix Factorization using Gradient Descent

Practice for Matrix Factorization Implementation

**Matrix Factorization Mathematical Derivation**

We know that the matrix factorization breaks the rating matrix $R$ into two matrices $U \in \mathbb{R}^{\#users \times K}$ and $M \in \mathbb{R}^{K \times \#items}$ where K represent the latent space dimensionality.

$R$ can then be approximated through the following equation:

$$
\mathbf{R} \approx \mathbf{U} \times \mathbf{M} = \hat{\mathbf{R}}
$$

The error, incorporating the regularization term, is calculated as follows:

$$
e_{ij}^2 = (r_{ij} - \sum_{k=1}^K{u_{ik}m_{kj}})^2 + \frac{\beta}{2} \sum_{k=1}^K{(||U||^2 + ||M||^2)}
$$

In order to be able to use Stochastic Gradient Descent (SGD) to optimize $U$ and $M$, we need to find the partial derivatives of the error function with respect to both $u_{ik}$ and $m_{kj}$. The partial derivatives can be derived as follows:

$$
\frac{\partial}{\partial u_{ik}}e_{ij}^2 = -2(r_{ij} - \hat{r}_{ij})(m_{kj}) + \frac{\beta}{2} \times (2 u_{ik}) = -2 e_{ij} m_{kj} + \beta u_{ik}
$$

$$
\frac{\partial}{\partial m_{ik}}e_{ij}^2 = -2(r_{ij} - \hat{r}_{ij})(u_{ik}) + \frac{\beta}{2} \times (2 m_{kj}) = -2 e_{ij} u_{ik} + \beta m_{kj}
$$

Thus the update rules will be:

$$
u'_{ik} = u_{ik} + \alpha \frac{\partial}{\partial u_{ik}}e_{ij}^2 = u_{ik} - \alpha(-2 e_{ij} m_{kj} + \beta u_{ik} ) = u_{ik} + \alpha(2 e_{ij} m_{kj} - \beta u_{ik} )
$$

$$
m'_{kj} = m_{kj} + \alpha \frac{\partial}{\partial m_{kj}}e_{ij}^2 = m_{kj} - \alpha(-2 e_{ij} u_{ik} + \beta m_{kj} ) = m_{kj} + \alpha(2 e_{ij} u_{ik} - \beta m_{kj} )
$$

## `0` Construct Utility Matrix from the Data

In [ ]:
utility_matrix = None
utility_matrix

songID  2263    2726    3785    8063    12709   13859   16548   17029   \
userID                                                                   
0          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5          0.0     0.0     0.0     2.0     0.0     0.0     0.0     0.0   
7          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
14         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
199976     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
199980     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
199988     0.0     5.0     0.0     0.0     0.0     0.0     0.0     0.0   
199990     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
199996     5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

songID  19299   19670   ...  113954  119103  120147  122065  123176  125557  \
userID                  ...                                                   
0          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
7          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
14         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
199976     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
199980     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
199988     5.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
199990     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
199996     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

songID  126757  131048  132189  134732  
userID                                  
0          0.0     0.0     0.0     0.0  
4          0.0     0.0     0.0     0.0  
5          0.0     0.0     0.0     0.0  
7          0.0     0.0     0.0     3.0  
14         0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
199976     5.0     0.0     0.0     0.0  
199980     0.0     0.0     0.0     0.0  
199988     0.0     0.0     0.0     0.0  
199990     0.0     0.0     0.0     0.0  
199996     0.0     0.0     0.0     0.0  

[53963 rows x 56 columns]

## `i` Matrices Initialization

Initialize the two random weights matrices $U$ and $M$ (Try $K=3$)

**Note**: Refer to the next cell for the dimensions of $U$ and $M$

**Hine**: You may use a function from `numpy.random` module (see, [Documentation](https://numpy.org/doc/stable/reference/random/index.html))

In [ ]:
U = None
M = None

## `ii` Define a Function to Implement Matrix Factorization

**Function Parameters**:
- `R`: Utility Matrix [of shape: (no. of users, no. of items)]
- `U`: User Latent Features Array [of shape: (no. of users, K)]
- `M`: Items Latent Features Array [of shape: (K, no. of items)]
- `epochs`: No. of training epochs
- `lr`: Learning rate (alpha)
- `beta`: Regularization Parameter

**Function Output**:
- `U`: Optimized User Latent Features Array
- `M`: Optimized Items Latent Features Array

**Main Procedures**:
1. Calculate predicted ratings
2. Calculate MSE Error
3. Calculate gradients
4. Update $U$ and $M$


In [ ]:
def matrix_factorization(R: np.ndarray, U: np.ndarray, M: np.ndarray, epochs: int, lr: float, beta: float):
    """
    Function Parameters:
    - `R`: Utility Matrix [of shape: `(no. of users, no. of items)`]
    - `U`: User Latent Features Array [of shape: `(no. of users, K)`]
    - `M`: Items Latent Features Array [of shape: `(K, no. of items)`]
    - `epochs`: No. of training epochs
    - `lr`: Learning rate (alpha)
    - `beta`: Regularization Parameter

    Function Output:
    - `U`: Optimized User Latent Features Array
    - `M`: Optimized Items Latent Features Array
    """

    # Confirm that no. of features is consistent between U and M
    assert U.shape[1] == M.shape[0], f'U and M must have consistent K. Found K={U.shape[1]} for U and K={M.shape[0]} for M'

    # Extract No. of Features (K)
    K = None

    # Define the Epochs loop
    for epoch in range(epochs):
        # Loop over every element in R
        for None: # Loop over each user
            for None: # Loop over each item
                if None: # Only proceed if the current interaction (i, j) is not missing
                    eij = None # Calculate the error in prediction
                    for None: # Loop over each latent features dimension
                        # Update Rules for both U and M:
                        U[i, k] = None
                        M[k, j] = None

        ## Error Calculation ##
        e_last = e if epoch > 0 else 100000000
        e = 0 # Initialize a variable to accumelate the errors
        for None: # Loop over each user
            for None: # Loop over each item
                if None: # Only proceed if the current interaction (i, j) is not missing
                                # since we only calculate the error for interactions having a ground truth value

                    first_part = None # calculate the first part of the error
                    second_part = 0 # Initialize a variable to accumelate the second part of the error

                    for None: # Loop over each latent features dimension
                        second_part += None

                    e += first_part + second_part # accumelate the error to the total error

        print(f'Epoch {epoch+1}/{epochs}: Total Error = {e}')

        if e < 0.001 or e_last-e < 0.001: # Stop if error is so small or improvement is not significant
            break

    return U, M


### `[Bonus]` Vectorized Error Calculation

Can the error calculation part be vectorized to get rid of for loops?

If you would like a challenge, try to redefine the function in the next cell with a vectorized error calculation.

## `iii` Use the Function to to Optimize the $U$ and $V$

In [ ]:
U, M = matrix_factorization(R=utility_matrix.values, U=U, M=M, epochs=None, lr=None, beta=None)

## `iv` Recommend top-K Songs

Recommend top-K ($K=5$) songs for user ($userID=199988$)

Note: Make sure to filter songs they already rated before

Top 5 Recommended Items for User 199988:
	- Top 1 Song: 122065 (Predicted Rating: 6.006507760135539)
	- Top 2 Song: 125557 (Predicted Rating: 5.819802641310125)
	- Top 3 Song: 52611 (Predicted Rating: 5.740148229067794)
	- Top 4 Song: 79622 (Predicted Rating: 5.701478248255747)
	- Top 5 Song: 71582 (Predicted Rating: 5.691077461653294)


------------------------

# `04` Matrix Factorization using SVD Algorithm

Practice for using `SVD` algorithm implementation from `scikit surprise` library.

## `i` Prepare the Data

- Load the Data into `surprise` Dataset
- Split data into train and test


## `ii` Model Initialization

Instantiate two models:
- Model with baselines (biases)
- Model without baselines

**Note**: Use `surprise.prediction_algorithms.matrix_factorization.SVD` (see, [Documentation](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#:~:text=surprise.prediction_algorithms.matrix_factorization.-,SVD,-(n_factors%3D)))

In [ ]:
# Biased Model


In [ ]:
# Non-Biased Model


## `iii` Fit each Model on Training Data

In [ ]:
# Biased Model


In [ ]:
# Non-Biased Model


## `iv` Test both Models on the Testing Data

Compare the errors of the two models using multiple error formulas.

**Note**: Refer to `surprise.accuracy` module (see, [Documentation](https://surprise.readthedocs.io/en/stable/accuracy.html))

In [ ]:
# Biased Model


In [ ]:
# Non-Biased Model


## `v` Recommend Top $10$ Songs for User $199988$

Is there a difference in recommended songs from the two models?

In [ ]:
# Biased Model


In [ ]:
# Non-Biased Model


----------------------------------------------

$$ Wish \space you \space all \space the \space best \space ♡ $$
$$ Mahmoud \space Shawqi $$